refactor: copilot project

daabr opened the pull request:
Fixes in general:
- Use Copilot terminology from Copilot settings page: "cancel seat" instead of "un/subscribe" and "un/register" (except in the metadata description - where brevity is important)

Fixes in `README.md`:
- Removed ":" from a header
- Removed obsolete project var `GITHUB_ORG`
- Removed obsolete warning at the bottom, after https://github.com/autokitteh/docs.autokitteh.com/pull/143

Fixes in `autokitteh.yaml`:
- Removed obsolete project var `GITHUB_ORG`
- Renamed project vars and default Slack channel to be more self-describing
- `IDLE_USAGE_THRESHOLD`: changed debug values (10m) to stated default (72h)

Replaced `helpers.py` with `users.py` (adapted from new Purrr, like `helpers.py`)

Fixes in `users.py` compared to `helpers.py`:
- "Helpers" is too generic, always use specific names
- Never return `None` when a function declares that it returns only `str`
- When calling Slack functions, rely only in `try` & `except SlackApiError`, checking `resp.get("ok", False)` is never reached in case of an error (unlike our old Starlark wrapper)
- `_slack_users` function: Python rewrite doesn't need recursion (unlike our old Starlark wrapper)

Fixes in `msg.json`:
- UX principle: whenever possible, the "yes" option should be "first" (on the left), and the "no" option should be second (on the right)

Fixes in `seats.py`:
- Redundant constants and globals: if used only once, embedded them where they're used
- Use new project var names from the manifest (more descriptive)
- Lumped GitHub globals together
- `prune_idle_seats` function is unnecessary - embedded in `find_idle_seats` with a new input boolean
- `t` timestamp: renamed to be more self-describing (`now`), but more importantly eset for each seat instead of using a single initial value (the loop *should* be quick, but may not be)
- `status` line too line, split into 2 lines (readability is usually more important than efficiency)
- Reversed `is_idle` to `is_active`: results in flatter code at the bottom of the loop (`if is_active: continue` instead of `if is_idle: ...`)
- `seat_dict`: renamed, don't specify types in variable names
- `seat_dict`: changed type from `dict[str, Any]` to `dict[str, str]` - it's our object, no need to mimic GitHub's nested object structure
- Replaced "engage seat" (which doesn't mean a lot?) with "interact with user" in function name and comments
- Removed test-only comment and code
- Subscribe: use `event_type` instead of `data.type`
- Use more modern Python `match-case` instead of `if-elif-else`
- `report` function: handle gracefully empty `SLACK_LOG_CHANNEL` to enable making this project var optional

Note: we don't really need `autokitteh.start`, we could use the same trick as in the Slack sample project (embed important values e.g. GitHub/Slack user IDs in the interactive message and get them when the user clicks a button), but I guess there's educational value in showcasing `autokitteh.start`.

Fixes in `triggers.py`:
- `on_schedule` function: print each seat on its own instead of the whole list - more readable
- `_get_logins` function: renamed arg, don't specify types in variable names

daabr committed:
refactor: copilot project

daabr committed:
cosmetic tweak

daabr committed:
fix 2 optional ruff warnings

daabr committed:
fix: can't build list JSON files

daabr commented on github_copilot_seats/users.py:
```diff
@@ -0,0 +1,69 @@
+        real_name = profile.get("real_name", "").lower()
+        normalized_name = profile.get("real_name_normalized", "").lower()
+        if github_name in (real_name, normalized_name):
+            return user.id
```
```suggestion
            return user.get("id", "")
```

daabr committed:
Update github_copilot_seats/users.py

pashafateev commented on github_copilot_seats/README.md:
```diff
@@ -1,29 +1,33 @@
+that have access to [Copilot](https://github.com/features/copilot). If any of
+them haven't used it in a preceding period of time, it automatically marks
+their seat for cancellation in the next billing cycle, and notifies them in a
+Slack DM.
```
```suggestion
This automation runs daily to list all the users in the GitHub organization who have access to [Copilot](https://github.com/features/copilot). If any of
them haven't used it in a preceding period of time, it marks
their seat for cancellation in the next billing cycle, and notifies them in a
Slack DM.
```

pashafateev replied:
I think "enumerates" makes it unnecessarily difficult to understand. And I took out "automatically" because I think it's redundant.

pashafateev commented on github_copilot_seats/autokitteh.yaml:
```diff
@@ -6,15 +6,13 @@ version: v1
-    - name: LOG_CHANNEL
-      value: log
+    - name: SLACK_LOG_CHANNEL
+      value: copilot-log
```
`SLACK_LOG_CHANNEL` being pre-populated in my PR is an undesired remnant of testing . Are you sure we want to set this for the user?

daabr replied:
Now that it's optional, I see no problem in making it empty by default.

pashafateev commented on github_copilot_seats/seats.py:
```diff
@@ -1,117 +1,116 @@
-slack = slack_client("slack_conn")
-
-org = github.get_organization(ORG_NAME)
+org = github.get_organization(os.getenv("github_conn__target_name"))
```
My concern is that we’re prioritizing brevity at the expense of clarity. There’s no existing documentation for `github_conn__target_name`, and I only heard about it last week. Its name isn’t very intuitive. Can a user deduce that `github.get_organization` expects the organization name? Probably. Is breaking this process into two distinct steps technically redundant? Yes. However, making something that isn’t immediately clear more explicit is worth adding an extra line, in my opinion.

daabr replied:
OK, brought back `GITHUB_ORG_NAME` const.

pashafateev commented on github_copilot_seats/seats.py:
```diff
@@ -1,117 +1,116 @@
 
-        comparison = ">=" if is_idle else "<"
-        status = f"{seat.assignee.login}: {t} - {seat.last_activity_at} = {delta} {comparison} {IDLE_USAGE_THRESHOLD} minutes"
+        comparison = "<" if is_active else ">="
```
Can you expand on your thought process behind switching from `is_idle` to `is_active`?

daabr replied:
From the PR description:
> Reversed `is_idle` to `is_active`: results in flatter code at the bottom of the loop

```
if is_active:
  continue

block of code...
```

Vs.

```
if is_idle:
  block of code...
```

pashafateev commented on github_copilot_seats/seats.py:
```diff
@@ -1,117 +1,116 @@
-        This is designed to run as a child workflow using:
-        autokitteh.start(loc="seats.py:engage_seat", data={"seat": seat})
+    Note - this function runs as a child workflow, by calling:
+    autokitteh.start(loc="filename.py:function_name", data={...})
```
style preference but I think the other version is easier to read and visually looks cleaner.

daabr replied:
Done.

daabr committed:
Update github_copilot_seats/README.md

Co-authored-by: Pasha Fateev <pasha@autokitteh.com>

daabr committed:
review comments

daabr committed:
ci: refresh settings and workflow (#146)

- Enable unit testing (except in the Purrr project - those UT require AK
context)
- Add various optional but important linters to Ruff
- Don't pin pytest and ruff versions, no point in updating it from time
to time like in the AK repo

daabr committed:
fix imports
